platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ExportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,208 @@
+import org.codice.ddf.migration.ExportMigrationContext;
+import org.codice.ddf.migration.ExportMigrationContextProxy;
+import org.codice.ddf.migration.ExportMigrationEntry;
+import org.codice.ddf.migration.MigrationEntry;
+import org.codice.ddf.migration.MigrationWarning;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class extends on the {@link ExportMigrationContext} interface to pre-create entries for
+ * all configuration objects.
+ */
+public class ExportMigrationConfigurationAdminContext extends ExportMigrationContextProxy {
+    private static final Logger LOGGER = LoggerFactory.getLogger(
+            ExportMigrationConfigurationAdminContext.class);
+
+    private final ConfigurationAdminMigratable admin;
+
+    private final Set<String> warnedExtensions = new HashSet<>(8);
+
+    private final Map<Path, ExportMigrationEntry> entries;
+
+    public ExportMigrationConfigurationAdminContext(ExportMigrationContext context,
+            ConfigurationAdminMigratable admin, Configuration[] configs) {
+        super(context);
+        Validate.notNull(admin, "invalid null configuration admin migratable");
+        Validate.notNull(configs, "invalid null configurations");
+        this.admin = admin;
+        this.entries = Stream.of(configs)
+                .filter(this::isValid)
+                .map(this::getEntry)
+                .collect(Collectors.toMap(MigrationEntry::getPath, Function.identity()));
+    }
+
+    /**
+     * This method is designed to circumvent a bug in Felix where it has been seen that a config object
+     * that is meant to represent a managed service factory is not properly created. Instead of being
+     * create using the createFactoryConfiguration(), it is obtained using getConfiguration(). After
+     * that, the proeprties are updated with the service factory pid as a property. Although it has
+     * that property, it ain't a real managed service factory as te factory or blueprint was never
+     * called to instantiate a corresponding service. In such case, we end up with a dummy config
+     * object in memory that is not attached to an actual instance of manager service.
+     * <p>
+     * If we were to export that object, we would end up re-creating it as a managed service factory
+     * on the other system and we potentially could end up with multiple instances of that manager service
+     * which could create problem. In addition, the system would not be a proper representation of the
+     * original system which didn't have that associated managed service.
+     * <p>
+     * We do not intent to protect against all possible stupidity they could do (e.g. different factory
+     * pid in properties than reported by the config object's interface). We only protect against
+     * the object not being created as a managed service factory when it should have been.
```

Comment:
❓ Is there a Felix ticket we can refer to here?

Reply:
I don't believe this is a Felix issue. Just that anybody can set properties on config objects and not create them properly. We saw Cellar doing that. It is just how the API is in Felix where one can create an config object by calling getConfiguration() which may re-create an object if it did not exist where the object is not a valid managed service factory. @brjeter anything else you want to add?

Reply:
I don't know that we've filed a ticket for every time we've seen it, but here's a related bug that I've created a ticket for: https://issues.apache.org/jira/browse/FELIX-5686. This also happened when we tried to register a second PersistenceManager incorrectly. There's no explicit issue currently in our system, but it is a common enough issue (whenever someone tries to call `getConfiguration()` for a managed service factory) that it may come up again. Hence the extra defense 😉